app/src/main/java/net/gsantner/markor/format/todotxt/TodoTxtTextActions.java

```diff
@@ -96,112 +94,78 @@ int getFormatActionsKey() {
-            final SttTaskWithParserInfo origTask = sttcmd.parseTask(origText, origSelectionStart);
             final CommonTextActions commonTextActions = new CommonTextActions(_activity, _hlEditor);
-
-            final Callback.a1<SttTaskWithParserInfo> cbUpdateOrigTask = (updatedTask) -> {
-                if (updatedTask != null) {
-                    SttCommander.SttTasksInTextRange rangeInfo = sttcmd.findTasksBetweenIndex(origText, origTask.getLineOffsetInText(), origTask.getLineOffsetInText());
-                    Editable editable = _hlEditor.getText();
-                    rangeInfo.startIndex = Math.max(rangeInfo.startIndex, 0);
-                    rangeInfo.endIndex = Math.max(rangeInfo.endIndex, 0);
-                    try {
-                        editable.delete(rangeInfo.startIndex, rangeInfo.endIndex);
-                    } catch (Exception e) {
-                        e.printStackTrace();
-                    }
-                    sttcmd.regenerateTaskLine(updatedTask);
-                    editable.insert(rangeInfo.startIndex, updatedTask.getTaskLine() + "\n");
-
-                    // Try to figure out new cursor pos
-                    int cursor = rangeInfo.startIndex + origTask.getCursorOffsetInLine();
-                    if (cursor != 0) {
-                        cursor += _hlEditor.getText().length() - origText.length(); // difference
-                    }
-                    if (cursor == _hlEditor.getText().length()) {
-                        cursor--; // Move to last char in text;
-                    }
-                    if (cursor >= 0 && cursor <= _hlEditor.getText().length()) {
-                        _hlEditor.setSelection(cursor);
-                    }
-                }
-            };
-
+            final TodoTxtTask[] selTasks = TodoTxtTask.getSelectedTasks(_hlEditor);
 
             switch (_action) {
                 case R.string.tmaid_todotxt_toggle_done: {
-                    origTask.setDone(!origTask.isDone());
+                    String replaceDone = "x ";
                     if (AppSettings.get().isTodoAddCompletionDateEnabled()) {
-                        origTask.setCompletionDate(SttCommander.getToday());
+                        replaceDone += TodoTxtTask.getToday() + " ";
                     }
-                    cbUpdateOrigTask.callback(origTask);
+                    runRegexReplaceAction(
+                            // If task starts with a priority, replace priority with 'x ...'
+                            new ReplacePattern(TodoTxtTask.PATTERN_PRIORITY_ANY, replaceDone),
+                            // else if task stars with 'x + (completion date)?' replace with "" (i.e. remove done)
+                            new ReplacePattern(TodoTxtTask.PATTERN_COMPLETION_DATE, ""),
+                            // else replace task start with 'x ...'
+                            new ReplacePattern("^", replaceDone)
```

Comment:
I have also added a comment to show how RegexReplaceAction is basically a sequence of `if` - `else if` - `else if`